File: examen_final/src/main/java/Cliente50.py
import threading
import math
import random
import re
import socket
import threading

sumaarray = [0] * 40

class TCPClient50:
    def __init__(self, ip, listener):
        self.servermsj = None
        self.SERVERIP = ip
        self.SERVERPORT = 4444
        self.mMessageListener = listener
        self.mRun = False
        self.out = None
        self.incoming = None

    def sendMessage(self, message):
        if self.out is not None:
            self.out.write(message + '\n')
            self.out.flush()

    def stopClient(self):
        self.mRun = False

    def run(self):
        self.mRun = True
        try:
            serverAddr = socket.gethostbyname(self.SERVERIP)
            print("TCP Client - C: Conectando...")
            clientSocket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
            clientSocket.connect((serverAddr, self.SERVERPORT))
            try:
                self.out = clientSocket.makefile('w')
                print("TCP Client - C: Sent.")
                print("TCP Client - C: Done.")
                self.incoming = clientSocket.makefile('r')
                while self.mRun:
                    self.servermsj = self.incoming.readline()
                    if self.servermsj is not None and self.mMessageListener is not None:
                        self.mMessageListener(self.servermsj)
                    self.servermsj = None
            except Exception as e:
                print("TCP - S: Error", e)
            finally:
                clientSocket.close()
        except Exception as e:
            print("TCP - C: Error", e)

class Cliente50:
    def __init__(self):
        self.sum = [0.0] * 40
        self.mTcpClient = None
        self.sc = None

    def iniciar(self):
        threading.Thread(target=self.runTcpClient).start()

        salir = "n"
        while salir != "s":
            salir = input()
            self.clienteEnvia(salir)

    def runTcpClient(self):
        self.mTcpClient = TCPClient50("127.0.0.1", self.clienteRecibe)
        self.mTcpClient.run()

    def clienteRecibe(self, llego):
        if "evalua" in llego.strip():

            pattern = re.compile(r"(\w+) (\d+) (.+)")
            matcher = pattern.match(llego)

            if matcher:
                parte1 = matcher.group(1)
                epoca = matcher.group(2)
                wNews = matcher.group(3)

                numerosString = wNews[1:-1]
                numerosArray = numerosString.split(",")
                wNew = [float(num.strip()) for num in numerosArray]

                self.procesar(int(epoca), wNew)
            else:
                print("No se encontró una coincidencia.")

    def clienteEnvia(self, envia):
        if self.mTcpClient is not None and envia != "s":
            self.mTcpClient.sendMessage(envia)

    def procesar(self, epoca, wNew):
        rna01s = []
        threads = []
        peso = []
        error = 9999

        for _ in range(6):
            rna01 = Rna01(7, 6, 1, epoca, wNew)
            rna01s.append(rna01)
            thread = threading.Thread(target=rna01.run)
            threads.append(thread)
            thread.start()

        try:
            for thread in threads:
                thread.join()

                if rna01.getError() < error:
                    error = rna01.getError()
                    peso = rna01.getPesos()
        except Exception as e:
            print(e)

        self.clienteEnvia(str(error) + " " + str(peso))


class Rna01:
    rand = random.Random()

    def __init__(self, ci, co, cs, epoca, wNew):
        self.ci = ci
        self.co = co
        self.cs = cs

        self.xin = []
        self.xout = []
        self.y = []
        self.s = []
        self.g = []
        self.w = []

        self.c = [0] * 3

        self.c[0] = ci
        self.c[1] = co
        self.c[2] = cs

        for _ in range(co + cs):
            self.y.append(0)
            self.s.append(0)
            self.g.append(0)

        for _ in range(ci * co + co * cs):
            if epoca == 1:
                self.w.append(self.getRandom())
            else:
                self.w.append(wNew.pop(0))

    def getRandom(self):
        return self.rand.uniform(-1, 1)

    def fun(self, d):
        return 1 / (1 + math.exp(-d))

    def entrenamiento(self, inp, out, veces):
        self.xin = inp
        self.xout = out

        for _ in range(veces):
            for i in range(len(self.xin)):
                self.entreno(i)

    def getPesos(self):
        return self.w

    def entreno(self, cii):
        ii = 0
        pls = 0
        ci = cii

        # entrenamiento

        # Ida
        # capa1
        ii = 0
        pls = 0

        for i in range(self.c[1]):
            for j in range(self.c[0]):
                pls += self.w[ii] * self.xin[ci][j]
                ii += 1

            self.s[i] = pls
            self.y[i] = self.fun(self.s[i])
            pls = 0

        # capa2
        pls = 0
        ii = self.c[0] * self.c[1]

        for i in range(self.c[2]):
            for j in range(self.c[1]):
                pls += self.w[ii] * self.y[j]
                ii += 1

            self.s[i + self.c[1]] = pls
            self.y[i + self.c[1]] = self.fun(self.s[i + self.c[1]])
            pls = 0

        # Vuelta
        # capa2 g
        for i in range(self.c[2]):
            self.g[i + self.c[1]] = (self.xout[ci][i] - self.y[i + self.c[1]]) * self.y[i + self.c[1]] * (
                    1 - self.y[i + self.c[1]])
            self.error = self.g[i + self.c[1]]

        # capa1 g
        pls = 0
        for i in range(self.c[1]):
            for j in range(self.c[2]):
                pls += self.w[self.c[0] * self.c[1] + j * self.c[1] + i] * self.g[self.c[1] + j]

            self.g[i] = self.y[i] * (1 - self.y[i]) * pls
            pls = 0

        # capa2 w
        ii = self.c[0] * self.c[1]
        for i in range(self.c[2]):
            for j in range(self.c[1]):
                self.w[ii] += self.g[i + self.c[1]] * self.y[j]
                ii += 1

        # capa1 w
        ii = 0
        for i in range(self.c[1]):
            for j in range(self.c[0]):
                self.w[ii] += self.g[i] * self.xin[ci][j]
                ii += 1

    def getError(self):
        return self.error

    def run(self):
        datos = [[0.5, 0.5, 0.0, 0.5, 1.0, 1.0, 1.0],
                 [0.5, 0.0, 0.0, 1.0, 1.0, 1.0, 0.5],
                 [0.5, 0.5, 0.0, 0.5, 0.0, 1.0, 0.0],
                 [0.5, 0.5, 0.0, 1.0, 0.0, 1.0, 1.0],
                 [0.0, 0.0, 0.0, 0.5, 0.0, 1.0, 0.0],
                 [0.0, 0.0, 1.0, 0.5, 1.0, 1.0, 1.0],
                 [0.0, 0.5, 0.0, 0.5, 0.0, 1.0, 1.0]]

        salida = [[0.5],
                  [0.5],
                  [1.0],
                  [0.5],
                  [1.0],
                  [0.5],
                  [1.0]]

        self.entrenamiento(datos, salida, 20)


# Ejemplo de uso
if __name__ == "__main__":
    objcli = Cliente50()
    objcli.iniciar()

File: examen_final/src/main/java/Cliente50.java
import java.util.Arrays;
import java.util.Scanner;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.Random;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
class Cliente50 {


    public double[] sum = new double[40];
    TCPClient50 mTcpClient;
    Scanner sc;

    public static void main(String[] args) {
        Cliente50 objcli = new Cliente50();
        objcli.iniciar();
    }

    void iniciar() {
        new Thread(
                () -> {
                    mTcpClient = new TCPClient50("127.0.0.1", this::ClienteRecibe);
                    mTcpClient.run();
                }
        ).start();
        //---------------------------

        String salir = "n";
        sc = new Scanner(System.in);

        while (!salir.equals("s")) {
            salir = sc.nextLine();
            ClienteEnvia(salir);
        }


    }

    void ClienteRecibe(String llego) {

        if (llego.trim().contains("evalua")) {

            Pattern pattern = Pattern.compile("(\\w+) (\\d+) (.+)");
            Matcher matcher = pattern.matcher(llego);

            if (matcher.matches()) {
                // Obtener las partes separadas
                String parte1 = matcher.group(1);
                String epoca = matcher.group(2);
                String wNews = matcher.group(3);

                String numerosString = wNews.substring(1, wNews.length() - 1);

                // Dividir la cadena en subcadenas individuales
                String[] numerosArray = numerosString.split(",");

                // Crear un nuevo arreglo de tipo double[]
                double[] wNew = new double[numerosArray.length];

                // Convertir las subcadenas en números double
                for (int i = 0; i < numerosArray.length; i++) {
                    wNew[i] = Double.parseDouble(numerosArray[i].trim());
                }
                procesar(Integer.parseInt(epoca),wNew);

            } else {
                System.out.println("No se encontró una coincidencia.");
            }


        }
    }


    void ClienteEnvia(String envia) {

        if (mTcpClient != null && !envia.equals("s")) {

            mTcpClient.sendMessage(envia);
        }
    }


    void procesar(int epoca,double[] wNew ) { // implementacion para 6 hilos

        Rna01[] rna01s = new Rna01[6];
        Thread[] threads = new Thread[6];
        double [] peso = new double[0];
        double error = 9999;
        for (int i = 0; i < 6; i++) {
            rna01s[i] = new Rna01(7,6,1,epoca, wNew);
            threads[i] = new Thread(rna01s[i]);
            threads[i].start();
        }
        try {
            for (int i = 0; i < 6; i++) {
                threads[i].join();
                if( rna01s[i].getError() < error){
                    error = rna01s[i].getError();
                    peso = rna01s[i].getPesos();
                }
            }
        } catch (InterruptedException e) {
            e.printStackTrace();
        }

        ClienteEnvia(error+" "+Arrays.toString(peso));

    }
}


class Rna01 implements Runnable{
    static final Random rand = new Random();
    int ci;
    int co;
    int cs;

    double xin[][];//={{0,1,0},{0,1,1},{1,0,0},{1,0,1}};
    //double xin[][]={{0,1,0},{0,1,1},{1,0,0},{1,0,1}};
    double xout[][];//={{1},{0},{1},{0}};

    double y[];

    //   double w[]={2,-2,0,1,3,-1,3,-2};
//   double s[]={0,0,0};
    double s[];
    double g[];
    //   double g[]={0,0,0};
    double w[];

    double error;

    //   int c[] = {3,2,1};//capas de datos
    int c[]=new int[3];//capas de datos

    public Rna01(int ci_,int co_,int cs_,int epoca, double [] wNew){

        this.ci=ci_;//2
        this.co=co_;//3
        this.cs=cs_;//1

        y = new double[co+cs];
        s = new double[co+cs];
        g = new double[co+cs];
        w = new double[ci*co+co*cs];

        c[0]=ci;
        c[1]=co;
        c[2]=cs;

        for(int i=0;i<y.length;i++){
            y[i]=0;s[i]=0;g[i]=0;
        }
        for(int i=0;i<w.length;i++){
            if(epoca==1) w[i]=getRandom();
            else w[i] = wNew[i];
        }


    }

    public double fun(double d){
        return 1/(1+Math.exp(-d));
    }

    public void printxingreso(){
        //visualizar x ingreso
        for(int i=0;i<xin.length;i++)
            for(int j=0;j<xin[i].length;j++)
                System.out.println("xingreso["+i+","+j+"]="+xin[i][j]);
        System.out.println("                ");
    }


    double getRandom() {
        return (rand.nextDouble() * 2 - 1); // [-1;1[
    }

    public void entrenamiento(double[][] in,double[][] sal,int veces){
        xin=in;
        xout=sal;
        for(int v=0;v<veces;v++)
            for(int i=0;i<xin.length;i++){
                entreno(i);
            }

    }
    public double[] getPesos(){
        return w;
    }
    public void entreno(int cii){
        int ii;
        double pls;
        int ci;

        //entrenamiento

        //////******** Ida**********//////
        //+++++++capa1
        ///ci=0;//entrenamiento primero   /////HOPE
        ci=cii;
        ii = 0;//capa0*capa1
        pls=0;
        for(int i=0;i<c[1];i++){
            for(int j=0;j<c[0];j++){
                pls=pls+w[ii]*xin[ci][j];
                ii++;
            }
            s[i]=pls;  //i = i+ capa0
            y[i]=fun(s[i]); //i = i+ capa0
            pls=0;
        }
        //++++++capa2
        pls=0;
        ii = c[0]*c[1];//capa1*capa2
        for(int i=0;i<c[2];i++){
            for(int j=0;j<c[1];j++){
                pls=pls+w[ii]*y[j];
                ii++;
            }
            s[i+c[1]]=pls;  //i = i + capa1
            y[i+c[1]]=fun(s[i+c[1]]); //i = i + capa1
            pls=0;
        }



        //////----------Fin Ida--------/////
        //////******** Vuelta**********/////
        //++++capa2 g
        for(int i=0;i<c[2];i++){
            g[i+c[1]]=(xout[ci][i]-y[i+c[1]])*y[i+c[1]]*(1-y[i+c[1]]);
            error = g[i+c[1]];
        }

        //++++capa1 g
        pls=0;
        for(int i=0;i<c[1];i++){
            for(int j=0;j<c[2];j++){
                pls=pls+w[c[0]*c[1]+j*c[1]+i]*g[c[1]+j];
            }
            g[i]=y[i]*(1-y[i])*pls;
            pls=0;
        }

        //++++capa2 w
        ii = c[0]*c[1];//capa1*capa2
        for(int i=0;i<c[2];i++){
            for(int j=0;j<c[1];j++){
                w[ii]=w[ii]+g[i+c[1]]*y[j];
                ii++;
            }
        }

        //++++capa1 w
        ii = 0;//capa0*capa1
        for(int i=0;i<c[1];i++){
            for(int j=0;j<c[0];j++){
                w[ii]=w[ii]+g[i]*xin[ci][j];
                ii++;
            }
        }

    }

    public double getError() {
        return error;
    }

    @Override
    public void run() {
            double[][] datos = new double[][]{{0.5,0.5,0.0,0.5,1.0,1.0,1.0},
                                                {0.5,0.0,0.0,1.0,1.0,1.0,0.5},
                                                {0.5,0.5,0.0,0.5,0.0,1.0,0.0},
                                                {0.5,0.5,0.0,1.0,0.0,1.0,1.0},
                                                {0.0,0.0,0.0,0.5,0.0,1.0,0.0},
                                                {0.0,0.0,1.0,0.5,1.0,1.0,1.0},
                                                {0.0,0.5,0.0,0.5,0.0,1.0,1.0}};

            double[][] salida = new double[][]{{0.5},
                                                {0.5},
                                                {1.0},
                                                {0.5},
                                                {1.0},
                                                {0.5},
                                                {1.0}};

           entrenamiento(datos,salida,20);

    }
}
